design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,327 @@
+
+#### Releases
+- **Alpha**:
+  An initial release of the feature for experimental purposes.
+  Recommended for non-production usages, evaluation or testing.
+
+  The API is considered unstable and may change significantly.
+  There are no backward compatability considerations and it can
+  be removed at any time.
+
+  The period in which a feature can remain in Alpha is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **Beta**:
+  The first release that can be evaluated with care in production.
+  Acting as a pre-release, its main objective is to collect feedback
+  from users to assure its usefulness and readiness for graduation.
+  If there is no confidence of usage or usefulness, it may remain in
+  this stage for some time.
+
+  However, the period in which a feature can remain in Beta is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The API is considered stable with care not to break backward compatibility
+  with previous beta releases.
+  This implies that fields may only be added during this stage,
+  not removed or renamed.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **GA**:
+  The feature graduated to general-availability (GA) and is now part of
+  the core features.
+
+  The API is considered stable with care not to break backward compatibility
+  with the previous releases.
+
+  The feature functionality is no longer controlled by a FG.
+
+> **Warning**: A Feature Gate flag is solely intended to control
+> feature lifecycle. It should not be confused and used as a cluster
+> configurable enablement of the functionality.
+> In cases where the cluster admin should control a functionality,
+> regardless to the feature stage, dedicated configuration field/s
+> should be included.
+
+#### Removal
+If a feature is targeted for deprecation and retirement,
+it needs to pass a deprecation process, depending on its current
+release stage (Alpha, Beta, GA).
+
+For more details, see [here](#deprecation-and-removal).
+
+#### Release Stage Transition Table
+The following table summarized the different release stages with their
+transition requirements and restrictions.
+
+| Stage          | Period range    | F.Gate | Removal Availability       |
+|----------------|-----------------|--------|----------------------------|
+| Alpha          | 1 to 2 releases | YES    | Between **minor** releases |
+| Beta           | 1 to 3 releases | YES    | Between **minor** releases |
+| GA             | -               | NO     | Between **major** releases |
```

Comment:
The CPUManager can be disabled afaik - just one example.

Reply:
Yes... I missed the `The feature is always enabled; you cannot disable it.` Not sure why some of the features can still be disabled that way.

I think the main issue for me is that, unlike k8s features, most of KubeVirt's features don't have a secondary config option to disable it.

Reply:
It's not a blocker from my side but an area of concern. We should be able to provide a way for admins to disable a functionality. One of the well known requests is for the admin to disable all VFIO related functionality, which isn't a single feature per se.

Reply:
@vladikr I definitely agree, but I think that we're mixing FGs and configuration which are intended for different purposes.

Enabling a feature gate basically means "as a user I understand there's a risk in enabling this feature - and I'm fine with it". It makes sense only when a feature is not GAed.

In addition, a feature (might) have APIs to enable/disable it, configure it, etc. The APIs to configure a feature can change drastically between features and need to be discussed on a case-to-case basis.

As an example, the k8s `NodeSwap` feature (which is currently Beta) has a feature gate. However, if you enable the feature gate, swap is still not accessible. This is because swap is configured with a "swap behavior" which defaults to `NoSwap`. So in order to actually use swap, the user would need to both enable the feature gate and actually enable the feature via its API/configurations. When `NodeSwap` would become GA, the feature gate will be enabled by default (and eventually removed), but there will still be configurations to enable/disable it.

Does that make sense to you?

Reply:
> We should be able to provide a way for admins to disable a functionality. One of the well known requests is for the admin to disable all VFIO related functionality, which isn't a single feature per se.

BTW this is a valid concern IMO.

I think this is an ongoing process and it's OK to have some exceptions, especially in the beginning. We should extra carefully discuss removing currently existing FGs since some of the features might need to be refactored in order to do that responsibly. Perhaps some features would need to re-examine their API design and add a knob to enable/disable (or other knobs) if necessary.

Reply:
@vladikr 
> We should be able to provide a way for admins to disable a functionality. One of the well known requests is for the admin to disable all VFIO related functionality, which isn't a single feature per se.

We can simply add cluster level configs to configure a feature

Reply:
> @vladikr
> 
> > We should be able to provide a way for admins to disable a functionality. One of the well known requests is for the admin to disable all VFIO related functionality, which isn't a single feature per se.
> 
> We can simply add cluster level configs to configure a feature.

@fabiand , So something like
```
featureGates:
    - featureA
    - featureB
...
```
? ;)

Reply:
:) Rather:
```yaml
featureGates:
    - featureA   # only required for Alpa + Beta, enforce from GA on
    - featureB
featureA:
  enabled: true  # used to disable the feature if needed.
  foo: 10
  bar: Baz
```

Reply:
I honestly don't see a reason to add a dedicated configuration section for each new feature that we graduate. It may result in a very large kubevirt cr - and will complicate things in the virt-config.
But don't see me as an obstacle in this matter.
@xpivarc do you have any thoughts on this?

Reply:
I would change our current features gate model to be similar to the k8s one. Then, we could provide the default values based on the graduation phases.
Once a feature graduated we can decide whether to remove the feature gate entirely or keep (those that make sense to keep)
This way we will avoid introducing API for features that don't require it.

Reply:
I do not think it is possible, or it will be a real challenge to do something like this.
The FG list is considered stable v1, to change it we will need to release v2 (and continue supporting v1).

But even if we do manage to change the API in v1, keeping FG enabled by default and using them after GA is not possible, at least it contradicts their meaning (see k8s ref in this long thread).
In addition, we have an interest to avoid having FG living in the project forever, something we suffer today with.

Lastly, the use of a general configuration knob has been already used in many occasions (e.g. VM rollout strategy) in addition to the FG that protects it.
Therefore, I think that for the common use-case, there need to add a special enable/disable knob to control a funtionality is rarely required. It is usually implicitly interpreted from other configuration knobs or by just simply not giving choices.

Reply:
> I do not think it is possible, or it will be a real challenge to do something like this. The FG list is considered stable v1, to change it we will need to release v2 (and continue supporting v1).

I'm sure that with a small degree of creativity, we can achieve this... ;) VMs in K8S were also "not possible".
> 
> But even if we do manage to change the API in v1, keeping FG enabled by default and using them after GA is not possible, at least it contradicts their meaning (see k8s ref in this long thread). In addition, we have an interest to avoid having FG living in the project forever, something we suffer today with.
> 

We don't have to mimic everything K8S does. We also don't have to keep *all* feature gates but only those that are needed, without introducing another API to disable the features.

> Lastly, the use of a general configuration knob has been already used in many occasions (e.g. VM rollout strategy) in addition to the FG that protects it. Therefore, I think that for the common use-case, the need to add a special enable/disable knob to control a functionality is rarely required. It is usually implicitly interpreted from other configuration knobs or by just simply not giving choices.

Again, the rollout strategy had a feature gate because we experimented with the API. 
I think KubeVirt needs to have a way to allow admins to disable features - even after GA and yes, not all features will need this.

Reply:
> @iholder101 maybe you can provide an APi example of how in the case of node swap the FG + config looks pre-ga and post-ga.

Sure:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration

# This section is for pre-Beta.
# This basically means "I understand that swap is dangerous and still wish to use it"
featureGates:
  NodeSwap: true

# this is the configuration part which will always remain, even long after GA.
memorySwap:
  swapBehavior: LimitedSwap # Default is "NoSwap".
```

As said above I agree that we need to separate between feature gates and enablement configs as they serve different purposes IMO.

Reply:
@vladikr I agree with you. Imho in Kubevirt when we talk about feature gates we talk about accepting risk and enabling certain configuration at the same time. We do not have different knobs for all features and there are features that just works if feature gate is here. On the other hand our user guide does suggest that our feature gates are equal to Kubernetes ones https://kubevirt.io/user-guide/operations/activating_feature_gates/.

I don't see conflict here as the features gates are still there and we do not say they are GA even thought we do believe there are such features. Therefore this discussion is about future, what we want to do, at least this is how I understand it. @EdDev @fabiand @iholder101 please keep me honest.

2.
I agree that it might be overwhelming to add new configs for all the features that currently has the feature gates. Imho nobody had a look at this and we don't know how many will be needed and what would be the best way to tackle this. Do we keep this in Kubevirt CR? Do we create per component configs as Kubernetes has? These are all the questions I am missing answers to, yet this lifecycle doesn't dictate one or the another. I think this is fine until we don't get to enforce/implement this design.

Reply:
Thanks @xpivarc.
I agree. As I said [here](https://github.com/kubevirt/community/pull/251#discussion_r1580706413), I think that this is an ongoing effort that will not end by the time this PR merges. I think we can all agree that existing features deserve a per-case discussion. However I do believe that in the future we should make the effort to differentiate between accepting a risk of an immature feature and a feature configuration.

As for the overwhelming argument, I think that eventually once a feature GAs, we can design the enablement configs to look something like:
```yaml
featureOne: FeatureOne{}
featureTwo: FeatureTwo{}
featureThree: FeatureThree{}
```

Which is essentially very similar to:
```yaml
featureGates:
- featureOne
- featureTwo
- featureThree
```

During the time a feature is not GAed the user would have to both enable a feature gate and the config itself. IMO this is not a bug but a feature, as the idea of taking an extra step in order to explicitly accept the risk is a good thing that helps protects the user. The burden of enabling in two places is acceptable for non-GAed features IMO.

Reply:
hmm. This document, once merged, is going to set our policy regarding feature lifecycle and their corresponding feature gates.
If we want to keep the decision about the feature gates vague I would expect this line to be removed `| GA             | -               | NO     | Between **major** releases |`
Otherwise, we should say that we expect the feature gates behavior to be discussed in the follow-up PR and not commit here to anything specific.

Reply:
Given all of the above I'd suggest replacing `NO` with `Optional` and moving this discussion into a new document/PR specific to FGs in KubeVirt to unblock this. Happy to help draft something in the coming weeks if folks agree.

Reply:
+1

Reply:
I understand the concerns raised here, but I feel we are in a deadlock.

The proposal is based on the need to stop queuing FGs indefinitely and with the understanding that the current observation of what a FG is in Kubevirt [1] is well known and understood.

What you are attempting to argue here is that the current FG definition needs to change, together with it API (that I honestly to not understand how you can change).
To me its sounds like another proposal.

It makes sense to defer things to a follow up discussion, but in this case, I do not understand how we can do it without breaking this whole thing.

The maintainers of the project have the power to use exceptions to keep FGs forever and redefining what it means in parallel. I may not agree with what you are trying to do, but at least when the FG is redefined, this document can be adjusted accordingly.

I will need help here to see how this suggestion here can be done without making all the rest of this proposal relevant to the goals set.

[1] https://kubevirt.io/user-guide/operations/activating_feature_gates/#activating-feature-gates

Reply:
A few notes:

- This proposal is about our future desired state.
- Existing FGs should be untouched by this proposal, but can be covered by follow up work. AKA we should not block on the current state
- FGs are about declaring a feature as stable and always available.
- *NOT* every feature requires a configuration. We can simply say that some feature (like LiveMigration) is turned on by default _without_ configuration (enable/disable). Optionally we can say that we want to add configuration to enable/disable it, or configure some detail.

IOW
- Removing - or having a FG unrevokably enabled - means that we as KubeVirt consider this feature to be useable for all.
- If we think that the feature should _not_ always be used, then _can_ add a configuration.

Decision diagram wise it would look like this:

```mermaid
flowchart TD
    Start --> QCFG{Requires configuration?}
    QCFG -- Yes --> AddConfig[Add configuration]
    QCFG -- No --> QDEFAULT
    AddConfig --> QDEFAULT

    QDEFAULT{Should be default?}
    QDEFAULT -- Yes --> MakeDefault[Make default] --> QOPTOUT
    QDEFAULT -- No --> QOPTIN

    QOPTIN{Requires Opt-In?} -- Yes --> AddOptIn[Add Opt-In] --> END
    QOPTOUT{Requires Opt-Out?} -- Yes --> AddOptOut[Add Opt-Out] --> END

    QOPTIN & QOPTOUT -- No --> END

    END
```

This should all be decided before GA. `Configuration` and `Opt-In`/`Opt-Out` can however even be added post-GA.

Reply:
> I understand the concerns raised here, but I feel we are in a deadlock.

Yup apologies by attempting to break the deadlock here I've just made it worse. /o\

My intent with the `Optional` suggestion was to defer making any concrete decisions on GA FGs until we also talk about reworking the current API but I see that's not helpful in moving this proposal forward.

Given the already included warning about FGs not replacing cluster configurables for features admins might need to configure and/or disable I'm personally fine keeping the requirement to drop _new_ feature FGs as they graduate to GA for now.

> @lyarwood @vladikr IIUIC then the friction arises from the question around _todays_ FGs, is this correct?

The behaviour and IMHO abuse (using FGs to control feature configuration and not just visibility in the cluster) of the current implementation yes.

> Assumgin Yes to my question above, then a few notes:
> 
> * This proposal is about our future desired state.

The future feature lifecycle given the current FG implementation yes.

> * Existing FGs should be untouched by this proposal, but can be covered by follow up work. AKA we should not block on the current state

Yes that's my understanding.

> * FGs are about declaring a feature as stable and always available.

They don't define a feature as stable or always available, quiet the opposite in fact given the current proposal.

> * _NOT_ every feature requires a configuration. We can simply say that some feature (like LiveMigration) is turned on by default _without_ configuration (enable/disable). Optionally we can say that we want to add configuration to enable/disable it, or configure some detail.
> 
> IOW
> 
> * Removing - or having a FG unrevokably enabled - means that we as KubeVirt consider this feature to be useable for all.
> * If we think that the feature should _not_ always be used, then _can_ add a configuration.

Yup agreed.

Reply:
> They don't define a feature as stable or always available, quiet the opposite in fact given the current proposal.

Yes, bad wording on my side in that regards. You are right :)

Thus in general agremeent, that's good, thanks for the quick feedback.
Let's wait for Vlaidk, and others.

Reply:
Following up on @vladikr 's concerns.
The main concern (Vladik, kepe me honest) is that there is a technical redundancy to toggle a feature.
A FG is technically allowing an admin to enable or disable a feature.
A enable/disable config item is also allowing an admin to enable or disable a feature.

Thus the question is: If we have the FG, why should we introduce a config toogle for the same functionality?

On the technical side there is not much that I have to say, maybe besides the fact that there are possibly slightly different nuances between "not present" and "disabled".

However, semantically I'd take the compromise of this technical redundancy in favor of a clear separation between
a) signaling a feature is stable (enabled by default, no FG)
b) admin is permitted to disable/enable certain features using a configuration

Thus even given the concerns, I'm advocating to move forward and keep this separation between feature life-cycle and feature configuration.